gc/base/SparseVirtualMemory.cpp

```diff
@@ -0,0 +1,235 @@
+	uintptr_t maxHeapSize = _heap->getMaximumMemorySize();
+	uintptr_t regionSize = _heap->getHeapRegionManager()->getRegionSize();
+	uintptr_t regionCount = in_heap_size / regionSize;
+
+	/* Ideally this should be ceilLog2, however this is the best alternative as ceilLog2 currently does not exist */
+	uintptr_t ceilLog2 = MM_Math::floorLog2(regionCount) + 1;
+
+	uintptr_t off_heap_size = (uintptr_t)((ceilLog2 * in_heap_size) / 2.0);
+#if !defined(J9ZOS390)
+	Assert_MM_true(regionSize % _pageSize == 0);
+#endif
+	bool success = MM_VirtualMemory::initialize(env, off_heap_size, NULL, NULL, 0, memoryCategory);
+
+	if (success) {
+		void *sparseHeapBase = getHeapBase();
+		_sparseDataPool = MM_SparseAddressOrderedFixedSizeDataPool::newInstance(env, sparseHeapBase, off_heap_size);
+		if ((NULL == _sparseDataPool) || omrthread_monitor_init_with_name(&_largeObjectVirtualMemoryMutex, 0, "SparseVirtualMemory::_largeObjectVirtualMemoryMutex")) {
+			success = false;
+		}
+	}
+
+	if (success) {
+		Trc_MM_SparseVirtualMemory_createSparseVirtualMemory_success((void*)_heap->getHeapTop(), (void*)_heap->getHeapBase(), (void *)in_heap_size, (void *)maxHeapSize, (void *)regionSize, regionCount, (void *)off_heap_size, _sparseDataPool);
+	} else {
+		Trc_MM_SparseVirtualMemory_createSparseVirtualMemory_failure((void*)_heap->getHeapTop(), (void*)_heap->getHeapBase(), (void *)in_heap_size, (void *)maxHeapSize, (void *)regionSize, regionCount, (void *)off_heap_size, _sparseDataPool);
+	}
+
+	return success;
+}
+
+void
+MM_SparseVirtualMemory::tearDown(MM_EnvironmentBase *env)
+{
+	if (NULL != _sparseDataPool) {
+		_sparseDataPool->kill(env);
+		_sparseDataPool = NULL;
+	}
+
+	if (NULL != _largeObjectVirtualMemoryMutex) {
+		omrthread_monitor_destroy(_largeObjectVirtualMemoryMutex);
+		_largeObjectVirtualMemoryMutex = (omrthread_monitor_t) NULL;
+	}
+
+	MM_VirtualMemory::tearDown(env);
+}
+
+void
+MM_SparseVirtualMemory::kill(MM_EnvironmentBase *env)
+{
+	tearDown(env);
+	env->getForge()->free(this);
+}
+
+bool
+MM_SparseVirtualMemory::updateSparseDataEntryAfterObjectHasMoved(void *dataPtr, void *objPtr)
+{
+	omrthread_monitor_enter(_largeObjectVirtualMemoryMutex);
+	bool ret = _sparseDataPool->updateSparseDataEntryAfterObjectHasMoved(dataPtr, objPtr);
+	omrthread_monitor_exit(_largeObjectVirtualMemoryMutex);
+	return ret;
+}
+
+void *
+MM_SparseVirtualMemory::allocateSparseFreeEntryAndMapToHeapObject(void *proxyObjPtr, uintptr_t size)
+{
+	/* Commiting and decommiting memory sizes must be multiple of pagesize */
+	uintptr_t adjustedSize = MM_Math::roundToCeiling(_pageSize, size);
+
+	omrthread_monitor_enter(_largeObjectVirtualMemoryMutex);
+	void *sparseHeapAddr = _sparseDataPool->findFreeListEntry(adjustedSize);
+	bool success = MM_VirtualMemory::commitMemory(sparseHeapAddr, adjustedSize);
+
+#if !(defined(LINUX) && defined(J9VM_ARCH_X86))
+	OMRZeroMemory(sparseHeapAddr, adjustedSize);
+#endif /* !(defined(LINUX) && defined(J9VM_ARCH_X86)) */
```

Comment:
You can add comment before this block this is ultra safe decision, I am sure we can significantly reduce list of platforms required zeroing after commit